### ISSUE TITLE
Inverse of a non-invertible matrix over the integers is a (non-integer) integer matrix

### ISSUE THREAD

**Issue by james-d-mitchell** [gap-system/gap#4884]
In the current GAP master branch I get the following, inverting a non-invertible integer `MatrixObj` results in a matrix that believes it is an integer matrix but whose entries are not integers. 

### Observed behaviour
~~~
gap> mat := Matrix(Integers, [[0, -3, 0, -2], [-1, 1, -1, 0],
> [0, 1, 0, 1], [0, 0, 2, 0]]);;
gap> InverseOp(mat);
<4x4-matrix over Integers>
gap> Display(last);
<4x4-matrix over Integers:
[[ -1, -1, -2, -1/2 ]
 [ -1, 0, -2, 0 ]
 [ 0, 0, 0, 1/2 ]
 [ 1, 0, 3, 0 ]
]>
~~~
### Expected behaviour
~~~
gap> mat := Matrix(Integers, [[0, -3, 0, -2], [-1, 1, -1, 0],
> [0, 1, 0, 1], [0, 0, 2, 0]]);;
gap> InverseOp(mat);
fail
~~~
or perhaps at a stretch:
~~~
gap> mat := Matrix(Integers, [[0, -3, 0, -2], [-1, 1, -1, 0],
> [0, 1, 0, 1], [0, 0, 2, 0]]);;
gap> InverseOp(mat);
<4x4-matrix over Rationals>
gap> Display(last);
<4x4-matrix over Rationals:
[[ -1, -1, -2, -1/2 ]
 [ -1, 0, -2, 0 ]
 [ 0, 0, 0, 1/2 ]
 [ 1, 0, 3, 0 ]
]>
~~~

**Comment by hulpke** [comment]
This is an interesting question: What should `Inverse` do for a matrix declared over a ring, which is invertible over the fraction field? And should `Matrix` check whether the given entries actually lie over the ring? @ThomasBreuer might know.

**Comment by ThomasBreuer** [comment]
Concerning the example `Matrix(Integers, [[1, 1], [2, E(8)]])`, this input must yield either an error or `fail`, and the documentation of `Matrix` must state what happens in such cases.
Currently the documentation of `Matrix` just states that the returned matrix is defined over the given base domain, and the documentation of `BaseDomain` refers to an introductory section that states that the matrix entries must either lie in the base domain or naturally embed into the base domain.
Thus `Matrix` must check this property, and perhaps something like `MatrixNC` should be provided, which omits these checks.

The first example is more subtle. Again, the documentation should be made precise about what shall happen. The role of the `BaseDomain` value of matrix objects seems to be not defined very restrictively. There is the following statement about the idea behind vector and matrix objects.

> We  want  to  be  able  to write (efficient) code that is independent of the actual  representation  (in  the  sense of GAP's representation filters, see Section 13.4) and preserves it.

If both `Integers` and `Rationals` are supported for a given representation filter of matrix objects then one could argue that it would be allowed to create inverses over different base domains.
On the other hand, this would be against the idea of getting efficient code. And there is already code (for example a `KroneckerProduct` method) that accepts only arguments over the same base domain.
The documentation must state what is allowed in general, for example what the sum and the product of two matrix objects over different base domains are.
My understanding is that operations such as `Inverse` should not try to create matrix objects with new (guessed) base domains but should signal an error or return `fail`; for the sake of consistency (since we can think of `Inverse` calling `Matrix` for creating its result), I think that the behaviour should be the same as for `Matrix(Integers, [[1, 1], [2, E(8)]])`.

**Comment by james-d-mitchell** [comment]
I agree with @ThomasBreuer, I think the "correct" behaviour would be for `Inverse` and `Matrix(Integers, [[1, 1], [2, E(8)]])` to return `fail` or give an error, and that there should be a "no check" version called `MatrixNC`.

**Comment by james-d-mitchell** [comment]
Seems completely reasonable to me @hulpke, it's what we did in `Semigroups` and it's the test failures there that indicated this issue to me when converting from our own matrices to `MatrixObj`s.

**Comment by fingolfin** [comment]
I agree with @ThomasBreuer and @james-d-mitchell regarding what *should* happen. I like the idea of `MatrixNC`. We may also need `NewMatrixNC` (note that `NewMatrix` is intended to be the low-level way to construct a matrix, while `Matrix` methods are intended to be implemented in terms of `NewMatrix` -- however, I don't think we make that clear in the docs right now (because it wasn't always like this).

**Comment by fingolfin** [comment]
Likewise `IdentityMatrix` is implemented in terms of `NewIdentityMatrix` etc. -- the idea here being that not every MatrixObj implementation should have to implement all variants of `IdentityMatrix` etc. , they only need one; and then we also offer defaults for e.g. `NewIdentityMatrix` which delegate to `NewMatrix`. So that a basic MatrixObj implementation should be able to get away with only implementing a single `NewMatrix` constructor (modulo bugs and omissions in our code for ensuring that, of course). All of this ought to be documented, though *sigh*

**Comment by fingolfin** [comment]
I submitted issue #4891 to remind us of this

**Comment by fingolfin** [comment]
This has been resolved on master (probably in PR #5153 or perhaps #5340).

```
gap> mat := Matrix(Integers, [[0, -3, 0, -2], [-1, 1, -1, 0],
> [0, 1, 0, 1], [0, 0, 2, 0]]);;
gap> InverseOp(mat);
Error, the elements in <list> must lie in <basedomain>
*[1] Error( "the elements in <list> must lie in <basedomain>" );
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:90
 [2] MakeIsPlistVectorRep( v![1], list, true )
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:319
 [3] Vector( list[i], t )
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:687
 [4] Matrix( n, Length( n ), M )
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:1180
<function "InverseMutable [ IsPlistMatrixRep ]">( <arguments> )
 called from read-eval loop at *stdin*:7
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk>
gap> Matrix(Integers, [[1, 1], [2, E(8)]]);
Error, the elements in <list> must lie in <basedomain>
*[1] Error( "the elements in <list> must lie in <basedomain>" );
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:90
 [2] MakeIsPlistVectorRep( basedomain, ShallowCopy( list ), true )
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:180
 [3] method( req1, req2, req3 )
   @ /Users/mhorn/Projekte/GAP/gap/lib/methsel.g:231
 [4] NewVector( filterVectors, basedomain, list[i] )
   @ /Users/mhorn/Projekte/GAP/gap/lib/matobjplist.gi:215
 [5] method( req1, req2, req3, req4 )
   @ /Users/mhorn/Projekte/GAP/gap/lib/methsel.g:254
...  at *stdin*:7
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk>
```
